torch/_inductor/ir.py

```diff
@@ -2337,23 +2337,32 @@ def realize_into(cls, src, dst):
         if isinstance(src, TensorBox):
             src = src.data
 
-        # We copy the contents of src into dst. In most cases this should
-        # be fused into a single kernel by the scheduler.
-        # NOTE: We cannot change src's layout to mutate dst directly as this
-        # would alias src to dst, which is not correct as further mutations to
-        # dst would effect users of src.
-        src.realize_hint()
-
-        src = Pointwise.create(
-            device=src.get_device(),
-            dtype=src.get_dtype(),
-            inner_fn=src.make_loader(),
-            ranges=[
-                V.graph.sizevars.guard_equals(a, b)
-                for a, b in zip(src.get_size(), dst.get_size())
-            ],
-        ).data
-        src.realize()
+        if src.is_user_of(dst.get_name()) and (
+            isinstance(src.data, Buffer)
+            and isinstance(src.get_layout(), FlexibleLayout)
+        ):
+            # Skip the copy when
+            # 1. If src is_user_of dst, we suppose it has been realize in previous
+            #    step of: V.graph.mark_buffer_mutated(dst.get_name())
+            # 2. The src must be FlexibleLayout, since it will be changed to MutationLayout with dst
+            need_copy = False
```

Comment:
It seems to me like there are two possibilities. Either
1. `src` has another user and so we must copy to preserve correctness
2. `src` has no users other than the copy and so we would expect it to be eliminated by fusion

Do you know which case is happening in llama and if the second, why isn't this buffer being removed?

Reply:
- For llama, Compare the generated code before and after the regression, we find after regression there are 16 extra memory copy with buffers below from `buf288` to `buf333`

<img width="797" alt="image" src="https://github.com/pytorch/pytorch/assets/53841472/0c097947-d4a3-4cfd-b856-1dbc5e12506f">


- Take `buf333` and `arg91_1` as example:
  - Before regression, there is no `buf333`. The result will directly write into  `arg91_1` .
  - After regression, the result will first write into `buf333`, then copy into `arg91_1`  which is a mutation of another buffer `buf334`.
 
<img width="782" alt="image" src="https://github.com/pytorch/pytorch/assets/53841472/0f176029-adf7-4615-a094-9e03e4347c46">

Reply:
As for why `buf333` doesn't eliminate in `dead_node_elimination` https://github.com/pytorch/pytorch/blob/3ec33957eb8d774946d79db9d8d051be2c2c88cf/torch/_inductor/scheduler.py#L1285

- `buf333` has one user node of `buf334`.
```
buf333: SchedulerNode(ComputedBuffer)
buf333.writes = [MemoryDep('buf333', c0, {c0: 16777216})]
buf333.unmet_dependencies = [MemoryDep('buf264', c1 - 512, {c0: 32, c1: 524288})]
buf333.met_dependencies = 
    [   MemoryDep('arg91_1', c0, {c0: 16777216}),
        MemoryDep('arg91_1', c1, {c0: 32, c1: 524288})]
buf333.users = [NodeUser(node=SchedulerNode(name='buf334'), can_inplace=True, is_weak=False)]
buf333.group.device = cpu
buf333.group.iteration = ((32, 1024, 512), ())
buf333.sizes = ([32, 1024, 512], [])
class buf333_loop_body:
....
```
- `buf334` has user node of `output`.
```
buf334: SchedulerNode(ComputedBuffer)
buf334.writes = [MemoryDep('buf334', c0, {c0: 16777216})]
buf334.unmet_dependencies = [MemoryDep('buf333', c0, {c0: 16777216}), WeakDep(name='buf267')]
buf334.met_dependencies = [StarDep(name='arg91_1')]
buf334.users = [NodeUser(node=OUTPUT, can_inplace=False, is_weak=False)]
buf334.group.device = cpu
buf334.group.iteration = ((16777216,), ())
buf334.sizes = ([16777216], [])
buf334.mutations = ['arg91_1']
class buf334_loop_body:
    var_ranges = {z0: 16777216}
    index0 = z0
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('buf333', get_index)
        get_index_1 = self.get_index('index0')
        store = ops.store('buf334', get_index_1, load, None)
        return store
```

Reply:
The reason why `buf333` and `buf334` don't be fused is `var_ranges` of `buf333` is `(32, 1024, 512)`, but `var_ranges` of `buf334` is `(16777216,)` it fails the `can_fuse` check here: https://github.com/pytorch/pytorch/blob/ddb0c265110cfd053009dee34cbc5151bed8cddc/torch/_inductor/codegen/cpp.py#L2802

- Here is `buf333`
```
buf333: SchedulerNode(ComputedBuffer)
buf333.writes = [MemoryDep('buf333', c0, {c0: 16777216})]
buf333.unmet_dependencies = [MemoryDep('buf264', c1 - 512, {c0: 32, c1: 524288})]
buf333.met_dependencies = 
    [   MemoryDep('arg91_1', c0, {c0: 16777216}),
        MemoryDep('arg91_1', c1, {c0: 32, c1: 524288})]
buf333.users = [NodeUser(node=SchedulerNode(name='buf334'), can_inplace=True, is_weak=False)]
buf333.group.device = cpu
buf333.group.iteration = ((32, 1024, 512), ())
buf333.sizes = ([32, 1024, 512], [])
class buf333_loop_body:
    var_ranges = {z0: 32, z1: 1024, z2: 512}
   ....
```

- Here is `buf334`
```
buf334: SchedulerNode(ComputedBuffer)
buf334.writes = [MemoryDep('buf334', c0, {c0: 16777216})]
buf334.unmet_dependencies = [MemoryDep('buf333', c0, {c0: 16777216}), WeakDep(name='buf267')]
buf334.met_dependencies = [StarDep(name='arg91_1')]
buf334.users = [NodeUser(node=OUTPUT, can_inplace=False, is_weak=False)]
buf334.group.device = cpu
buf334.group.iteration = ((16777216,), ())
buf334.sizes = ([16777216], [])
buf334.mutations = ['arg91_1']
class buf334_loop_body:
    var_ranges = {z0: 16777216}
    ....
```

Reply:
In that case I think there are two ways forward, we either need to:
1. Prove that the buffer has no users and skip the copy
2. Enable cpu inductor to make this type of fusion

I think 1 should be possible by adding a special IR node for `Copy` and adding a copy-removal pass in the scheduler. Perhaps you could comment on 2?

Reply:
Thanks for the comment. Some questions I am not understanding clearly:

- _Actually in this case the buffer also has other users_ 
Looks like here `buf333` has only one user of `buf334`. And `buf334` has only one user of `Output`.  What's the `other users` discussed here?

- _aliasing is the only way to completely remove the redundant stores_
Did you mean making `buf334` of `buf333`'s alias like `buf334.data.layout = MutationLayout(buf333)`